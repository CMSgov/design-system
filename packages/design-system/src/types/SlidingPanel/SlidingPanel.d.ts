// disabling lint rules until component files are converted to TS
/* eslint-disable react/prefer-stateless-function */
/* eslint-disable filenames/match-exported */
import * as React from 'react';

export type SlidingPanelHeadingLevel = '1' | '2' | '3' | '4' | '5';

export interface SlidingPanelProps {
  /**
   * Gives more context to screen readers on the SlidingPanel close button.
   */
  ariaLabel?: string;
  closeButtonText?: React.ReactNode;
  children: React.ReactNode;
  className?: string;
  footerBody?: React.ReactNode;
  footerTitle?: string;
  /**
   * Enables focus trap functionality within SlidingPanel.
   */
  hasFocusTrap?: boolean;
  /**
   * Text for the SlidingPanel heading. Required because the `heading` will be focused on mount.
   */
  heading?: React.ReactNode;
  /**
   * A unique `id` to be used on heading element to label multiple instances of SlidingPanel.
   */
  headingId?: string;
  /**
   * Heading type to override default `<h3>`
   */
  headingLevel?: SlidingPanelHeadingLevel;
  /**
   * Enables "sticky" position of SlidingPanel header element.
   */
  isHeaderSticky?: boolean;
  /**
   * Enables "sticky" position of SlidingPanel footer element.
   */
  isFooterSticky?: boolean;
  onCloseClick: (...args: any[]) => any;
}

export default class SlidingPanel extends React.Component<SlidingPanelProps, any> {
  render(): JSX.Element;
}
