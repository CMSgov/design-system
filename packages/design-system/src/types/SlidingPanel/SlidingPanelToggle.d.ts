// disabling lint rules until component files are converted to TS
/* eslint-disable react/prefer-stateless-function */
/* eslint-disable filenames/match-exported */
import * as React from 'react';

export interface SlidingPanelToggleProps {
  /**
   * Determines if SlidingPanel is open or closed.
   * This value is used to re-focus the toggle that opened the panel when the panel closes.
   */
  panelOpen: boolean;
  /**
   * SlidingPanelToggle content.
   */
  children: React.ReactNode;
  /**
   * Additional classes for the toggle button anchor element.
   */
  className?: string;
  /**
   * Adds `display: inline` to the SlidingPanelToggle.
   */
  inline?: boolean;
  /**
   * This function is called with an id that the toggle generates.
   * It can be used in implementing the SlidingPanel for keeping track of which panel the toggle controls.
   */
  showPanel: (...args: any[]) => any;
}

export default class SlidingPanelToggle extends React.Component<
  React.ComponentPropsWithRef<'button'> & SlidingPanelToggleProps,
  any
> {
  render(): JSX.Element;
}
