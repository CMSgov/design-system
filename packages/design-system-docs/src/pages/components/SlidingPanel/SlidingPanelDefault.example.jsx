/* eslint no-alert: 0 */
import { SlidingPanel, SlidingPanelToggle } from '@design-system';
import React from 'react';
import ReactDOM from 'react-dom';

class SlidingPanelDefaultExample extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      showPanel: false,
    };
  }

  togglePanel() {
    this.setState({ showPanel: !this.state.showPanel });
  }

  render() {
    return (
      <div>
        <p>
          <strong>Click the link below to open the default Sliding Panel.</strong>
        </p>

        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt
          ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation
          ullamco laboris nisi ut aliquip ex ea commodo consequat.
        </p>

        <SlidingPanelToggle panelOpen={this.state.showPanel} showPanel={() => this.togglePanel()}>
          Toggle a default help drawer.
        </SlidingPanelToggle>

        {this.state.showPanel && (
          <SlidingPanel
            footerTitle="Footer Title"
            footerBody={<p className="ds-text ds-u-margin--0">Footer content</p>}
            heading="Sliding Panel Heading"
            onCloseClick={() => this.togglePanel()}
          >
            <strong>This is a default Sliding Panel component</strong>

            <p>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor
              incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud
              exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure
              dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
              Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt
              mollit anim id est laborum.
            </p>
          </SlidingPanel>
        )}

        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt
          ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation
          ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in
          reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur
          sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id
          est laborum.
        </p>
      </div>
    );
  }
}

ReactDOM.render(<SlidingPanelDefaultExample />, document.getElementById('js-example'));
