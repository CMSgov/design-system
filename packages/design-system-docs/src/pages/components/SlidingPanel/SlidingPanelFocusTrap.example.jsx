/* eslint no-alert: 0 */
import { Button, SlidingPanel, SlidingPanelToggle } from '@design-system';
import React from 'react';
import ReactDOM from 'react-dom';

class SlidingPanelDefaultExample extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      showPanel: false,
    };
  }

  togglePanel() {
    this.setState({ showPanel: !this.state.showPanel });
  }

  render() {
    return (
      <div>
        <p>
          <strong>Click the link below to open a Sliding Panel with focus trap enabled.</strong>
        </p>

        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt
          ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation
          ullamco laboris nisi ut aliquip ex ea commodo consequat.
        </p>

        <SlidingPanelToggle panelOpen={this.state.showPanel} showPanel={() => this.togglePanel()}>
          Toggle a sliding panel with focus trap.
        </SlidingPanelToggle>

        {this.state.showPanel && (
          <SlidingPanel
            hasFocusTrap
            footerTitle="Footer Title"
            footerBody={<p className="ds-text ds-u-margin--0">Footer content</p>}
            heading="Sliding Panel Heading"
            onCloseClick={() => this.togglePanel()}
          >
            <strong>This is a Sliding Panel component with focus trap enabled.</strong>

            <p>
              To see focus trap in action, use the <kbd>TAB</kbd> key to navigate within the Sliding
              Panel. If the property <code>hasFocusTrap</code> is <code>true</code>, focus will not
              leave the Sliding Panel; if the property is set to <code>false</code>, focus will
              leave the Sliding Panel.
            </p>

            <p>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor
              incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud
              exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure
              dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
              Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt
              mollit anim id est laborum.
            </p>

            <Button>Inside Button (Does Nothing)</Button>
          </SlidingPanel>
        )}

        <Button className="ds-u-margin-top--2">Outside Button (Does Nothing)</Button>

        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt
          ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation
          ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in
          reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur
          sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id
          est laborum.
        </p>
      </div>
    );
  }
}

ReactDOM.render(<SlidingPanelDefaultExample />, document.getElementById('js-example'));
