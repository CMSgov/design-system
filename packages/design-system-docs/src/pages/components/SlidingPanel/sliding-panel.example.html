<div>
  <p>
    <strong>
      Note: This is just an example of the HTML markup. See the React example for a functional
      example.
    </strong>
  </p>

  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut
    labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco
    laboris nisi ut aliquip ex ea commodo consequat.
  </p>

  <button class="ds-c-button ds-c-button--transparent ds-c-sliding-panel__toggle" type="button">
    Toggle the help drawer.
  </button>

  <div
    aria-labelledby="slidingPanel_123"
    aria-modal="false"
    class="ds-c-sliding-panel"
    role="dialog"
  >
    <div class="ds-c-sliding-panel__window">
      <div class="ds-c-sliding-panel__header">
        <h3 tabindex="0" id="slidingPanel_123" class="ds-c-sliding-panel__header-heading">
          Help Drawer Heading
        </h3>

        <button
          aria-label="Close panel"
          class="ds-c-button ds-c-button--secondary ds-c-button--small ds-c-sliding-panel__close-button"
          type="button"
        >
          Close
        </button>
      </div>

      <div class="ds-c-sliding-panel__body">
        <strong>An Explanation</strong>

        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt
          ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation
          ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in
          reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur
          sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id
          est laborum.
        </p>
      </div>

      <div class="ds-c-sliding-panel__footer">
        <h4 class="ds-text ds-c-sliding-panel__footer-title">Footer title</h4>
        <div class="ds-text ds-c-sliding-panel__footer-body">Footer content</div>
      </div>
    </div>
  </div>

  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut
    labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco
    laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in
    voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat
    non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
</div>
