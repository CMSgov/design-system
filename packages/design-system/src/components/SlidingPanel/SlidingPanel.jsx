import FocusTrap from 'focus-trap-react';
import Button from '../Button/Button';
import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';
import uniqueId from 'lodash/uniqueId';

export class SlidingPanel extends React.PureComponent {
  constructor(props) {
    super(props);
    this.headingRef = null;
    this.id = this.props.headingId || uniqueId('slidingPanel_');
    this.handleEscapeKey = this.handleEscapeKey.bind(this);
  }

  componentDidMount() {
    if (this.props.hasFocusTrap) document.addEventListener('keydown', this.handleEscapeKey);
    if (this.headingRef) this.headingRef.focus();
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.handleEscapeKey);
  }

  handleEscapeKey(evt) {
    switch (evt.code) {
      case 'Escape':
        this.props.onCloseClick();
        break;
      default:
        break;
    }
  }

  render() {
    const {
      ariaLabel,
      children,
      className,
      closeButtonText,
      footerBody,
      footerTitle,
      hasFocusTrap,
      heading,
      isHeaderSticky,
      isFooterSticky,
      onCloseClick,
    } = this.props;
    const Heading = `h${this.props.headingLevel}` || `h3`;

    /* eslint-disable jsx-a11y/no-noninteractive-tabindex, react/no-danger */
    const slidingPanelMarkup = () => (
      <div
        aria-labelledby={this.id}
        className={classNames(className, 'ds-c-sliding-panel')}
        role="dialog"
      >
        <div className="ds-c-sliding-panel__window">
          <div className="ds-c-sliding-panel__header">
            <Heading
              tabIndex="0"
              id={this.id}
              className="ds-c-sliding-panel__header-heading"
              ref={(el) => (this.headingRef = el)}
            >
              {heading}
            </Heading>
            <Button
              aria-label={ariaLabel}
              className="ds-c-sliding-panel__close-button"
              size="small"
              onClick={onCloseClick}
            >
              {closeButtonText}
            </Button>
          </div>
          <div
            className={classNames('ds-c-sliding-panel__body', {
              'ds-c-sliding-panel--is-sticky': isHeaderSticky || isFooterSticky,
            })}
          >
            {children}
          </div>
          <div className="ds-c-sliding-panel__footer">
            <h4 className="ds-c-sliding-panel__footer-title">{footerTitle}</h4>
            <div className="ds-c-sliding-panel__footer-body">{footerBody}</div>
          </div>
        </div>
      </div>
    );

    return (
      <>
        {hasFocusTrap ? (
          <FocusTrap focusTrapOptions={{ clickOutsideDeactivates: true }}>
            {slidingPanelMarkup()}
          </FocusTrap>
        ) : (
          slidingPanelMarkup()
        )}
      </>
    );
  }
}

SlidingPanel.defaultProps = {
  ariaLabel: 'Close help drawer',
  closeButtonText: 'Close',
  headingLevel: '3',
};

// TODO: closeButtonText, heading should be a string, but it is being used as a node in MCT,
// until we provide a better solution for customization, we type it as a node.
SlidingPanel.propTypes = {
  /**
   * Gives more context to screen readers on the SlidingPanel close button.
   */
  ariaLabel: PropTypes.string,
  closeButtonText: PropTypes.node,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  footerBody: PropTypes.node,
  footerTitle: PropTypes.string,
  /**
   * Enables focus trap functionality within SlidingPanel.
   */
  hasFocusTrap: PropTypes.bool,
  /**
   * Text for the SlidingPanel heading. Required because the `heading` will be focused on mount.
   */
  heading: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
  /**
   * A unique `id` to be used on heading element to label multiple instances of SlidingPanel.
   */
  headingId: PropTypes.string,
  /**
   * Heading type to override default `<h3>`
   */
  headingLevel: PropTypes.oneOf(['1', '2', '3', '4', '5']),
  /**
   * Enables "sticky" position of SlidingPanel header element.
   */
  isHeaderSticky: PropTypes.bool,
  /**
   * Enables "sticky" position of SlidingPanel footer element.
   */
  isFooterSticky: PropTypes.bool,
  onCloseClick: PropTypes.func.isRequired,
};

export default SlidingPanel;
