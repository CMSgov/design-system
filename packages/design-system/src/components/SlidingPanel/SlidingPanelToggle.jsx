import Button from '../Button/Button';
import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';

/**
 * A link that triggers the visibility of a sliding panel
 */
class SlidingPanelToggle extends React.PureComponent {
  componentDidUpdate(prevProps) {
    if (!this.props.panelOpen && prevProps.panelOpen && this.buttonRef) {
      this.buttonRef.focus();
    }
  }

  render() {
    const { className, children, inline, showPanel, panelOpen, ...others } = this.props;
    const classes = classNames(
      'ds-c-sliding-panel__toggle',
      inline && 'ds-c-sliding-panel__toggle--inline',
      className
    );

    return (
      <Button
        className={classes}
        inputRef={(el) => (this.buttonRef = el)}
        onClick={showPanel}
        variation="transparent"
        {...others}
      >
        {children}
      </Button>
    );
  }
}

SlidingPanelToggle.propTypes = {
  /**
   * Determines if SlidingPanel is open or closed.
   * This value is used to re-focus the toggle that opened the panel when the panel closes.
   */
  panelOpen: PropTypes.bool.isRequired,
  /**
   * SlidingPanelToggle content.
   */
  children: PropTypes.node.isRequired,
  /**
   * Additional classes for the toggle button anchor element.
   */
  className: PropTypes.string,
  /**
   * Adds `display: inline` to the SlidingPanelToggle.
   */
  inline: PropTypes.bool,
  /**
   * This function is called with an id that the toggle generates.
   * It can be used in implementing the SlidingPanel for keeping track of which panel the toggle controls.
   */
  showPanel: PropTypes.func.isRequired,
};

export default SlidingPanelToggle;
