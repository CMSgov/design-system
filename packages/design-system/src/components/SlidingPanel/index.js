export { default as SlidingPanel } from './SlidingPanel';
export { default as SlidingPanelToggle } from './SlidingPanelToggle';
