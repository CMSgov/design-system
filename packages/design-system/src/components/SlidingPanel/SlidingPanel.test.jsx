import SlidingPanel from './SlidingPanel.jsx';
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

const defaultProps = {
  footerBody: (
    <div>
      <p>Some footer content</p>
    </div>
  ),
  footerTitle: 'Footer title',
  onCloseClick: () => {},
  heading: 'Sliding panel title',
};

function renderSlidingPanel(props) {
  props = Object.assign({}, defaultProps, props);
  const wrapper = shallow(
    <SlidingPanel {...props}>
      <p>content</p>
    </SlidingPanel>
  );
  return { props, wrapper };
}

describe('SlidingPanel', () => {
  it('calls props.onCloseClick on close button click', () => {
    const onCloseClick = jest.fn();
    const { wrapper } = renderSlidingPanel({ onCloseClick });
    const closeBtn = wrapper.find('Button');
    closeBtn.simulate('click');
    expect(onCloseClick).toHaveBeenCalled();
  });

  it('renders a snapshot', () => {
    const tree = renderer
      .create(
        <SlidingPanel {...defaultProps}>
          <p>content</p>
        </SlidingPanel>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
