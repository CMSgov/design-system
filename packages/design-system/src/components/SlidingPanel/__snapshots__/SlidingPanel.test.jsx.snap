// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlidingPanel renders a snapshot 1`] = `
<div
  aria-labelledby="slidingPanel_2"
  className="ds-c-sliding-panel"
  role="dialog"
>
  <div
    className="ds-c-sliding-panel__window"
  >
    <div
      className="ds-c-sliding-panel__header"
    >
      <h3
        className="ds-c-sliding-panel__header-heading"
        id="slidingPanel_2"
        tabIndex="0"
      >
        Sliding panel title
      </h3>
      <button
        aria-label="Close help drawer"
        className="ds-c-button ds-c-button--small ds-c-sliding-panel__close-button"
        onClick={[Function]}
        type="button"
      >
        Close
      </button>
    </div>
    <div
      className="ds-c-sliding-panel__body"
    >
      <p>
        content
      </p>
    </div>
    <div
      className="ds-c-sliding-panel__footer"
    >
      <h4
        className="ds-c-sliding-panel__footer-title"
      >
        Footer title
      </h4>
      <div
        className="ds-c-sliding-panel__footer-body"
      >
        <div>
          <p>
            Some footer content
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
